filament/src/MaterialInstanceManager.h

```diff
@@ -0,0 +1,84 @@
+/*
+ * Copyright (C) 2025 The Android Open Source Project
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <utils/bitset.h>
+
+#include <vector>
+
+namespace filament {
+
+class FMaterial;
+class FMaterialInstance;
+class FEngine;
+
+// This class manages a cache of material instances mapped by materials. Having a cache allows us to
+// re-use instances across frames.
+class MaterialInstanceManager {
+public:
+    class Record {
+    public:
+        Record(FMaterial const* material)
+            : mMaterial(material) {}
+        ~Record() = default;
+
+        Record(Record const& rhs) noexcept = default;
+        Record& operator=(Record const& rhs) noexcept = default;
+        Record(Record&& rhs) noexcept = default;
+        Record& operator=(Record&& rhs) noexcept = default;
+
+        void terminate(FEngine& engine);
+        void reset() { mAvailable = utils::bitset32{ (1 << (mInstances.size() + 1)) - 1 }; }
+        FMaterialInstance* getInstance();
+
+    private:
+        FMaterial const* mMaterial = nullptr;
+        std::vector<FMaterialInstance*> mInstances;
+        utils::bitset32 mAvailable = {};
+
+        friend class MaterialInstanceManager;
+    };
+
+    ~MaterialInstanceManager() = default;
```

Comment:
Add a ctor and copy/move ctor and = operators (just to be clean). Implement them in the .ccp file to they're not inlined.